tests/sync/session/progress_notifications.cpp

```diff
@@ -24,7 +24,128 @@
+    auto session = sync_session_with_bind_handler(server, user, "/test-sync-progress-stale-flag",
+                                                  [&](auto&, auto&, std::shared_ptr<SyncSession>) {
+                                                      bind_handler_called = true;
+                                                  },
+                                                  [](auto, auto) { },
+                                                  SyncSessionStopPolicy::Immediately);
+    EventLoop::main().run_until([&] { return bind_handler_called == true && SyncSession::OnlyForTesting::has_stale_progress(*session); });
+    REQUIRE(!session->is_in_error_state());
+    REQUIRE(!SyncSession::OnlyForTesting::has_fresh_progress(*session));
+
+    std::atomic<int> call_count(0);
+    std::atomic<uint64_t> transferred(0);
+    std::atomic<uint64_t> transferrable(0);
+    uint64_t current_transferred = 0;
+    uint64_t current_transferrable = 0;
+
+    SECTION("streaming notifiers run with or without stale data") {
+        session->register_progress_notifier([&](auto xferred, auto xferable) {
+            transferred = xferred;
+            transferrable = xferable;
+            ++call_count;
+        }, SyncSession::NotifierType::download, true);
+        // Wait for the initial callback.
+        EventLoop::main().run_until([&] { return call_count.load() > 0; });
+        
+        // Send a stale notification
+        call_count = 0;
+        current_transferred = 12;
+        current_transferrable = 202;
+        SyncSession::OnlyForTesting::handle_progress_update(*session, current_transferred, current_transferrable, 3, 103, false);
+        // Require that the notifier finally be called.
+        EventLoop::main().run_until([&] { return call_count.load() > 0; });
+        CHECK(transferred == current_transferred);
+        CHECK(transferrable == current_transferrable);
+
+        // Send a fresh notification
+        call_count = 0;
+        current_transferred = 13;
+        current_transferrable = 203;
+        SyncSession::OnlyForTesting::handle_progress_update(*session, current_transferred, current_transferrable, 3, 103, true);
+        // Require that the notifier finally be called.
+        EventLoop::main().run_until([&] { return call_count.load() > 0; });
+        CHECK(transferred == current_transferred);
+        CHECK(transferrable == current_transferrable);
+    }
+
+    SECTION("non-streaming notifiers don't run until fresh data is available") {
+        session->register_progress_notifier([&](auto xferred, auto xferable) {
+            transferred = xferred;
+            transferrable = xferable;
+            ++call_count;
+        }, SyncSession::NotifierType::download, false);
+
+        // Send a stale notification
+        SyncSession::OnlyForTesting::handle_progress_update(*session, 10, 200, 1, 100, false);
+        std::atomic<int> spin_count(0);
```

Comment:
There's no need for this to be atomic, since it can only ever be updated from the current thread. The state we update from within sync callbacks needs to be atomic since those callbacks run on the sync client's thread.